### ISSUE TITLE
Create TODO.md

### ISSUE THREAD

**Issue by AKushWarrior** [isar/isar#4]
If you want to post a TODO.md, I'd be happy to put some time into this and the other Isar libraries.

I presume that Isar is meant to be "Hive 2.0", essentially Dart bindings to lmdb?

**Comment by AKushWarrior** [comment]
That's right. I'll think about that one then :).

**Comment by frank06** [comment]
Happy to see you @leisim in open action again! 🎉 

> So for now I decided against relationships and backlinks but if many mobile developers request this feature, we might add it in the future.

If this is of any help, I see how you could focus on core features (like #2 - fantastic!) and let third party libraries implement higher level features?

Fwiw, I implemented [relationships in Flutter Data](https://flutterdata.dev/data-support/relationships/) using a global bi-directed graph notifier. It works with Hive as local storage, but without using Hive built-in relationships.

I'm really looking forward to having Isar as default backend.

**Comment by AKushWarrior** [comment]
> If this is of any help, I see how you could focus on core features (like #2 - fantastic!) and let third party libraries implement higher level features?
> 
> Fwiw, I implemented [relationships in Flutter Data](https://flutterdata.dev/data-support/relationships/) using a global bi-directed graph notifier. It works with Hive as local storage, but without using Hive built-in relationships.

The problem is that this is still inefficient on web because IndexedDB has NO parallelization whatsoever. Thus, providing relationships would only be helpful on native, which goes against the stated purpose of cross platform feature parity.

On the other hand, I fully agree that some features (ex. foreign database integration drivers) should be abstracted into other libraries.

**Comment by frank06** [comment]
Interesting, I never worked with IndexedDB. So if I had a to-many relationship modeled using a foreign key, issuing a query with a where clause on that PK is inefficient?

**Comment by AKushWarrior** [comment]
> Interesting, I never worked with IndexedDB. So if I had a to-many relationship modeled using a foreign key, issuing a query with a where clause on that PK is inefficient?

I had a diagram somewhere I sketched out, of sample relations and how internal architecture affects them. Let me see if I can dig it up.

**Comment by AKushWarrior** [comment]
Okay I found the diagrams, but they're more or less incomprehensible. Some notes I pulled off them:
- There are two purposes to relationships: maintaining referential integrity and performance improvements.
   * Referential integrity improvements don't really make sense in the context of a NoSQL database.
   * It's impossible to implement relationships (AFAIK) in a manner beyond the naive:
      - "take a user inputted object (one of the many) which contains a foreign key"
      - "go find the correspondent object to that foreign key and then store it with the user inputted object"
   * This method is not great; it essentially balloons space usage. The issue is that we can't hold a memory pointer or reference to an object in another database in IndexedDB. In retrospect, this is obvious; the memory address of a given object in IndexedDB varies from instance to instance, even on the same machine.
